awx/main/tasks/system.py

```diff
@@ -720,45 +721,43 @@ def handle_work_success(task_actual):
 
 
 @task(queue=get_local_queuename)
-def handle_work_error(task_id, *args, **kwargs):
-    subtasks = kwargs.get('subtasks', None)
-    logger.debug('Executing error task id %s, subtasks: %s' % (task_id, str(subtasks)))
-    first_instance = None
-    first_instance_type = ''
-    if subtasks is not None:
-        for each_task in subtasks:
-            try:
-                instance = UnifiedJob.get_instance_by_type(each_task['type'], each_task['id'])
-                if not instance:
-                    # Unknown task type
-                    logger.warning("Unknown task type: {}".format(each_task['type']))
-                    continue
-            except ObjectDoesNotExist:
-                logger.warning('Missing {} `{}` in error callback.'.format(each_task['type'], each_task['id']))
-                continue
+def handle_work_error(task_actual):
+    try:
+        instance = UnifiedJob.get_instance_by_type(task_actual['type'], task_actual['id'])
+    except ObjectDoesNotExist:
+        logger.warning('Missing {} `{}` in error callback.'.format(task_actual['type'], task_actual['id']))
+        return
+    if not instance:
+        return
 
-            if first_instance is None:
-                first_instance = instance
-                first_instance_type = each_task['type']
-
-            if instance.celery_task_id != task_id and not instance.cancel_flag and not instance.status in ('successful', 'failed'):
-                instance.status = 'failed'
-                instance.failed = True
-                if not instance.job_explanation:
-                    instance.job_explanation = 'Previous Task Failed: {"job_type": "%s", "job_name": "%s", "job_id": "%s"}' % (
-                        first_instance_type,
-                        first_instance.name,
-                        first_instance.id,
-                    )
-                instance.save()
-                instance.websocket_emit_status("failed")
+    subtasks = instance.get_jobs_fail_chain()  # reverse of dependent_jobs mostly
+    logger.debug(f'Executing error task id {task_actual["id"]}, subtasks: {[subtask.id for subtask in subtasks]}')
+
+    deps_of_deps = {}
+
+    for subtask in subtasks:
+        if subtask.celery_task_id != instance.celery_task_id and not subtask.cancel_flag and not subtask.status in ('successful', 'failed'):
+            # If there are multiple in the dependency chain, A->B->C, and this was called for A, blame B for clarity
+            blame_job = deps_of_deps.get(subtask.id, instance)
```

Comment:
@fosterseth this fixes the test case that we have today, for the case where there is a project update which is needed by an inventory update, which is needed by a job.

But the whole thing doesn't actually make any sense when you look hard enough. If you have that situation, the inventory update and the job both exist and are in a holding pattern, blocked by the `TaskManager`. When the task manager tries to advance these pending jobs, it runs through their `task.dependent_jobs.all()` which is... basically... the reverse of `get_job_fail_chain()` which is used here. The task manager then attaches that message

> Previous Task Failed:

And it does so in a way indistinguishable from here.

So, which code has responsibility for marking dependent jobs as failed (the job and inventory update in this case)? It seems the answer is "both"... er... whichever gets to it first. This is not a good idea.

The test case we had seems to have been working by accident, because the `dependent_jobs` fail chain changes over time, and the `subtasks` passed to the dispatcher for starting the original update was empty. After the task manager submitted the update, the inventory update was launched, and linked the update as a dependency. Because of that, this task did nothing, which deferred to the task manager, which applied the expected message. Had the specific means of launching been different (like putting the inventory update in a workflow), it would have been unlikely to give the expected error message.

Those are exactly the bigger-picture separation-of-responsibilities problems that https://github.com/ansible/awx/pull/13061 was concerned with. I just wanted to take this change to articulate one thing.